### PR TITLE
Close keyboard when clicking in the background

### LINKED ISSUE #6016: Close keyboard when clicking in the background
### Describe your suggested feature

Activities do not support closing the keyboard when clicking in the background, having this feature will def make a better user experience

https://user-images.githubusercontent.com/75211982/135745971-3210d09e-83ff-4966-afcb-7a52cef69165.mp4

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with an extension, I should be opening an issue in the [extensions repository](https://github.com/tachiyomiorg/tachiyomi-extensions/issues/new/choose).
- [X] I have updated the app to version **[0.12.2](https://github.com/tachiyomiorg/tachiyomi/releases/latest)**.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by cdhiraj40**
fixes #6016 

## Video

https://user-images.githubusercontent.com/75211982/135834001-49002de4-7eef-4d72-8c20-431450141bfb.mp4

### CODE REVIEW

**Comment by cdhiraj40** [comment]
hey @arkon I'm done with the work you can check :)

**Comment by ztimms73** [comment]
There is no need to ping. Sooner or later your pull request will be checked.
If the keyboard is hiding, then the focus with SearchView should disappear.

File: app/src/main/java/eu/kanade/tachiyomi/data/database/models/Manga.kt [hunk 1]
```diff
@@ -37,8 +37,7 @@ interface Manga : SManga {
 
     fun getGenres(): List<String>? {
         if (genre.isNullOrBlank()) return null
-        return genre?.split(", ")?.map { it.trim() } ?.filterNot { it.isBlank() } ?.distinct()
-        
+        return genre?.split(", ")?.map { it.trim() }?.filterNot { it.isBlank() }?.distinct()
```
Comment (ztimms73) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/Manga.kt:40]: Going beyond the name of the pull request. It is unclear why this was done.

Comment (cdhiraj40) [reply app/src/main/java/eu/kanade/tachiyomi/data/database/models/Manga.kt:40]: yes, actually I really don't know how it got changed, in the next commit I will make it as the way it was earlier.

**Comment by cdhiraj40** [comment]
> There is no need to ping a maintainer. Sooner or later your pull request will be checked. If the keyboard is hiding, then the focus from SearchView should disappear.

so then the cursor should disappear?

**Comment by cdhiraj40** [comment]
> There is no need to ping a maintainer. Sooner or later your pull request will be checked. If the keyboard is hiding, then the focus from SearchView should disappear.

I have got a better implementation that can solve this, will commit soon. Thanks for telling me :)

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt [hunk 2]
```diff
@@ -71,4 +75,21 @@ abstract class BaseThemedActivity : AppCompatActivity() {
             return resIds
         }
     }
+
+    // this  lets keyboard close when clicked in backgroud
+    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
+        if (event.action === MotionEvent.ACTION_DOWN) {
```
Comment (ivaniskandar) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:81]: Structural equality is enough

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt [hunk 3]
```diff
@@ -71,4 +75,21 @@ abstract class BaseThemedActivity : AppCompatActivity() {
...
+            val v = currentFocus
+            if (v is EditText) {
+                val outRect = Rect()
+                v.getGlobalVisibleRect(outRect)
+                if (!outRect.contains(event.rawX.toInt(), event.rawY.toInt())) {
+                    v.clearFocus()
+                    val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
```
Comment (ivaniskandar) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:88]: Use extension function `getSystemService<InputMethodManager>()`

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt [hunk 4]
```diff
@@ -71,4 +75,21 @@ abstract class BaseThemedActivity : AppCompatActivity() {
...
+                    imm.hideSoftInputFromWindow(v.getWindowToken(), 0)
```
Comment (ivaniskandar) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:89]: Property access syntax is available `v.windowToken`

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt (hunk 2 above)
Comment (cdhiraj40) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:81]: ```event.action == MotionEvent.ACTION_DOWN``` ?

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt (hunk 4 above)
Comment (cdhiraj40) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:89]: ```  val imm = getSystemService<InputMethodManager>() ``` ?

Comment (ivaniskandar) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:89]: You can use null-safe call here `getSystemService<InputMethodManager>()?.hideSoftInputFromWindow(v.windowToken, 0)`

Comment (cdhiraj40) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:89]: done and working perfectly fine I will commit the changes

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt [hunk 5]
```diff
@@ -71,4 +76,20 @@ abstract class BaseThemedActivity : AppCompatActivity() {
...
+    // this  lets keyboard close when clicked in backgroud
+    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
+        if (event.action == MotionEvent.ACTION_DOWN) {
+            val v = currentFocus
+            if (v is EditText) {
+                val outRect = Rect()
+                v.getGlobalVisibleRect(outRect)
+                if (!outRect.contains(event.rawX.toInt(), event.rawY.toInt())) {
+                    v.clearFocus()
+                    val imm = getSystemService<InputMethodManager>()?.hideSoftInputFromWindow(v.windowToken, 0)
```
Comment (FourTOne5) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:89]: getSystemService stuff are generally stored in system util at /app/src/main/java/eu/kanade/tachiyomi/util/system/ContextExtensions.kt

Comment (cdhiraj40) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/activity/BaseThemedActivity.kt:89]: This is not working/will work as the code which I have written has to been given to a parent activity which lets all fragments or child activity to hide the keyboard.

**Comment by arkon** [comment]
Minor issue: when tapping the "x" clear icon, the keyboard temporarily gets hidden but then pops back up when the text field takes focus again.

**Comment by cdhiraj40** [comment]
> Minor issue: when tapping the "x" clear icon, the keyboard temporarily gets hidden but then pops back up when the text field takes focus again.

I have tried with a  new approach but now it's only doing that when there is a list that is getting refreshed,
will look into and if you have any suggestions then please do share :)

**Comment by paulo27ms** [comment]
I reckon the keyboard should stay as the x is just to clear the text quickly, so same behavior as before your changes.

**Comment by arkon** [comment]
We don't use fragments. There's controllers that you may be able to use (e.g. [LibraryController](https://github.com/tachiyomiorg/tachiyomi/blob/master/app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryController.kt)), although I'm not entirely sure what you need to do with them.

**Comment by arkon** [comment]
Please open another PR if needed.

**Pull request closed without merging**